syntax = "proto3";

package analysis;

service AnalysisEngine {
  // Legacy method (File-based)
  rpc AnalyzeVideo(VideoRequest) returns (stream VideoResponse);

  // New Streaming method (Bi-directional)
  rpc StreamAnalysis(stream VideoChunk) returns (stream MetricsUpdate);
}

message VideoRequest {
  string video_path = 1;
  string calibration_path = 2;
  float confidence_threshold = 3;
  string match_id = 4;
  string model_path = 5;
}

message VideoResponse {
  string job_id = 1;
  string status = 2;
  float progress = 3;
  string message = 4;
  AnalysisResult result = 5;
}

message AnalysisResult {
  string match_id = 1;
  int32 total_frames = 2;
  int32 players_tracked = 3;
  string report_id = 4;
  string player_metrics_csv_path = 5;
  string ball_metrics_csv_path = 6;
}

// --- Streaming Messages ---

message VideoChunk {
  bytes data = 1;
  string match_id = 2;
  string calibration_path = 3;
  string model_path = 4;
  int32 chunk_index = 5;
  bool is_last_chunk = 6;
}

message MetricsUpdate {
  string status = 1;
  float progress = 2;
  string message = 3;
  repeated PlayerMetric metrics = 4;
  BallMetric ball_metric = 5;
}

message PlayerMetric {
  int32 player_id = 1;
  float x = 2;
  float y = 3;
  float speed = 4;
  string team_id = 5;
  int32 frame_index = 6;
}

message BallMetric {
  float x = 1;
  float y = 2;
  float z = 3; // For 3D projection if available
  bool is_possessed = 4;
  int32 frame_index = 5;
}
